src/peer-store/README.md

```diff
@@ -1,3 +1,156 @@
+
+---- (should be removed / re-written, but important for design decisions)
+
+One aspect that we need to consider is wether we should add information to every book, even if we don't have any relevant information for it. For instance, if we just discover a peer via a typical discovery service, we will have the `peerId` and an array of `multiaddr`. When we do `peerStore.set()`, should we also do `protoBook.set()` with an empty list of protocols? I don't see any advantage on adding to the remaining ones.
+
+**IMPORTANT:** This is one of the biggest design decisions to make (set vs events). The programmatic API is the easiest solution but it can provide users an API that they sould not need. If we go on an event-based approach, the `peerStore` should receive all the relevant subsystems (discovery, identifyService, ...) and sense the environment (via events) to gather the information that would need to be sent via the API. Thile the latest seems the best solution, it is the more complex one to implement, as we would ideally have an interface that those subsystems would need to implement and each time we have a new subsystem that needs to add data to the peerStore, we might need to update the `peer-store` codebase (or have a good set of abstractions).
+
+It is also important pointing out that users would be able to use `peerStore.protoBook.*`, so eventually we should move those into `peerStore._ protoBook.*` if we do not intend them to use it.
+
+---
+
+- `peerStore.get(peerId, options)`
+
+Get the information of a provided peer. The amount of information that we want can be customized with the following options, which are true by default:
+
+```js
+{
+  address: true,
+  proto: true,
+  key: true,
+  metadata: true
+}
+```
+
+- `peerStore.delete(peerId, [data])`
+
+Deletes the provided peer from every book. If data is provided, just remove the data from the books. The data should be provided as follows:
+
+```js
+{
+  address: [],
+  proto: [],
+  key: [],
+  metadata: []
+}
+```
+
+- `peerStore.peers(options)`
+
+Get an array of all the peers, as well as their information. The information intended can be customized with the following options, which are true by default:
+
+```js
+{
+  address: true,
+  proto: true,
+  key: true,
+  metadata: true
+}
+```
+
+## Address Book
+
+The `addressBook` keeps the known multiaddrs of a peer. The multiaddrs of each peer are not a constant and the Address book must have this into consideration.
+
+`Map<string, multiaddrInfo>`
+
+A `peerId.toString()` identifier mapping to a `multiaddrInfo` object, which should have the following structure:
+
+```js
+{
+  multiaddr: ,
+  validity: ,
+  confidence: 
+}
+```
+
+**Note:** except for multiaddr namings, the other properties are placeholders for now and might not be as described in the future milestones.
+
+- `addressBook.set()`
+- `addressBook.get()`
+- `getMultiaddrsForPeer()`
+- `addressBook.has()`
+- `addressBook.delete()`
+- `addressBook.peers()`
```

Comment:
Would `addressBook.peers` ever return different peers than `peerStore.peers`? If so, what value do we get from this difference? If not, should we remove this from the books and have users get it from PeerStore instead?

Reply:
the `peerStore.peers` would return a vision of things through the properties provided, `addr`, `protocols`, while the `addressBook.peers` would get the ones from `addr`. If we get rid of the combined methods from peerStore, should we also just provided the `addressBook.peers`, `protoBook.peers`?

Reply:
I'm not sure I am following. It would be helpful to see the API params and return value since this isn't stubbed in the code yet. I'm not clear on what this is supposed to do and why I wouldn't just use the PeerStore peers accessor.

Reply:
what would you expect to return from `peerStore.peers()`?

My initial idea was to do `peerStore.peers({ protocols, addresses })` and it would return all the peers we know with the "required" information. This would do internally `addressBook.peers()` + `protoBook.peers()` and assemble the data together. However, if we are dropping the other  `peerStore` API methods (like set, get, ...) and do atomic operations, I think we should also just support the underlying `peers()` method, which would return an array with all the map content.

Reply:
I expect `.peers` to give me a list of all the peers. What you're describing to me sounds like a `find` method which I do think is useful. Right now `peers` is just a Map property, it's an accessor, and I think it should stay an accessor. We should add find methods for searching.